File: internal/ir/ir.go
// Пакет ir — логічне (мова-агностик) представлення алгоритму як структури
// керування. Сюди парсер зводить будь-який код (Python/C++/…), а layout
// перетворює це на геометрію. Жодного рендера тут — лише структура.
package ir

// Node — елемент алгоритму. Реалізації: Process, IO, If (далі — While тощо).
type Node interface{ node() }

// Process — дія/обчислення (прямокутник ДСТУ).
type Process struct{ Text string }

func (*Process) node() {}

// IO — ввід/вивід (паралелограм ДСТУ).
type IO struct{ Text string }

func (*IO) node() {}

// If — розгалуження (ромб ДСТУ) з гілками «Так»/«Ні».
type If struct {
	Cond string
	Then *Block
	Else *Block
}

func (*If) node() {}

// For — цикл з лічильником (шестикутник ДСТУ «початок циклу»). Spec — підпис
// у форматі «змінна = початок, кінець, крок» (напр. «i = 0, n-1, 1»).
type For struct {
	Spec string
	Body *Block
}

func (*For) node() {}

// While — цикл з передумовою: ромб згори, тіло, дуга повернення.
type While struct {
	Cond string
	Body *Block
}

func (*While) node() {}

// Block — послідовність вузлів (виконуються згори вниз).
type Block struct{ Stmts []Node }

func (*Block) node() {}

// NewBlock — зручний конструктор послідовності.
func NewBlock(stmts ...Node) *Block { return &Block{Stmts: stmts} }

File: packages/engine/src/ir.ts
// IR — логічне (мова-агностик) дерево керування. Порт pkg/ir. Layout споживає це.
// Дискримінований union за полем kind. else/body/then завжди присутні як Block
// (порожній, якщо гілки нема) — на відміну від Go *Block (nil), щоб уникнути null-перевірок.

export interface Process { kind: 'process'; text: string; }
export interface IO { kind: 'io'; text: string; }
export interface Call { kind: 'call'; text: string; }
export interface Terminal { kind: 'terminal'; text: string; }
export interface If { kind: 'if'; cond: string; then: Block; else: Block; }
export interface For { kind: 'for'; spec: string; body: Block; else: Block; }
export interface While { kind: 'while'; cond: string; body: Block; else: Block; }
export interface DoWhile { kind: 'dowhile'; cond: string; body: Block; }
export interface InfLoop { kind: 'infloop'; body: Block; }
export interface Break { kind: 'break'; }
export interface Continue { kind: 'continue'; }
export interface Connector { kind: 'connector'; text: string; jump?: boolean; }
export interface Block { kind: 'block'; stmts: Node[]; }

export type Node =
  | Process | IO | Call | Terminal
  | If | For | While | DoWhile | InfLoop
  | Break | Continue | Connector | Block;

// Func — іменована програма (функція або тіло модуля) → окрема схема.
export interface Func { name: string; body: Block; main?: boolean; }
